dicom_utils/container/record.py

```diff
@@ -0,0 +1,242 @@
+    r"""Data structure for storing critical information about a DICOM file.
+    File IO operations on DICOMs can be expensive, so this class collects all
+    required information in a single pass to avoid repeated file opening.
+    """
+    path: Path
+    StudyInstanceUID: Optional[StudyUID]
+    SeriesInstanceUID: Optional[SeriesUID]
+    SOPInstanceUID: Optional[SeriesUID]
+
+    TransferSyntaxUID: Optional[TSUID]
+
+    Rows: Optional[int] = None
+    Columns: Optional[int] = None
+    NumberOfFrames: Optional[int] = None
+    PhotometricInterpretation: Optional[PI] = None
+    SimpleImageType: Optional[SIT] = None
+    ManufacturerModelName: Optional[str] = None
+    SeriesDescription: Optional[str] = None
+
+    PatientName: Optional[str] = None
+    PatientID: Optional[str] = None
+
+    def __hash__(self) -> int:
+        return hash(self.path)
+
+    def __eq__(self, other: Any) -> bool:
+        return isinstance(other, FileRecord) and self.path == other.path
+
+    @property
+    def is_image(self) -> bool:
+        return bool(self.Rows and self.Columns and self.PhotometricInterpretation)
+
+    @property
+    def is_volume(self) -> bool:
+        return self.is_image and ((self.NumberOfFrames or 1) > 1)
+
+    @property
+    def file_size(self) -> int:
+        return self.path.stat().st_size
+
+    @classmethod
+    def create(cls, path: Path) -> "FileRecord":
+        if not path.is_file():
+            raise FileNotFoundError(path)
+
+        with pydicom.dcmread(path, stop_before_pixels=True, specific_tags=tags) as dcm:
+            values = {tag.name: getattr(dcm, tag.name, None) for tag in tags}
+            for key in ("Rows", "Columns", "NumberOfFrames"):
+                values[key] = int(values[key]) if values[key] else None
+            img_type = ImageType.from_dicom(dcm).to_simple_image_type()
+            values["TransferSyntaxUID"] = dcm.file_meta.get("TransferSyntaxUID", None)
+
+        values.pop("ImageType")
+        return cls(path, SimpleImageType=img_type, **values)
+
+
+def record_iterator(
+    files: Sequence[Path], jobs: Optional[int] = None, use_bar: bool = True, threads: bool = False
+) -> Iterator[FileRecord]:
+    r"""Produces :class:`FileRecord` instances by iterating over an input list of files. If a
+    :class:`FileRecord` cannot be created from a path, it will be ignored.
+
+    Args:
+        files:
+            List of paths to iterate over
+
+        jobs:
+            Number of parallel processes to use
+
+        use_bar:
+            If ``False``, don't show a tqdm progress bar
+
+        threads:
+            If ``True``, use a :class:`ThreadPoolExecutor`. Otherwise, use a :class:`ProcessPoolExecutor`
+
+    Returns:
+        Iterator of :class:`FileRecord`s
+    """
+    files = list(p for p in files if p.is_file())
+    bar = tqdm(desc="Scanning files", total=len(files), unit="file", disable=(not use_bar))
+
+    def callback(f):
+        bar.update(1)
+
+    futures = []
```

Comment:
I think this can be cleaned up a little bit by adding a list comprehension and removing the `zip`:
```python
Pool = ThreadPoolExecutor if threads else ProcessPoolExecutor
with Pool(jobs) as p:
    futures = [p.submit(FileRecord.create, path) for path in files]
    for f in futures:
        f.add_done_callback(callback)

    for f in futures:
        if f.exception():
            continue
        elif record := f.result():
            yield record
```